src/Util/Api.php

```diff
@@ -187,30 +187,15 @@ public function completeQuery( $params ) {
 	 * @return PromiseInterface promise with the content of a page
 	 */
 	public function getPageAsync( $title ) {
-		return $this->queryAsync( [
-			'titles' => $title,
-			'prop' => 'revisions',
-			'rvprop' => 'content',
-			'rvparse' => true
-		] )->then( function ( array $result ) {
-			return $this->parseGetPageResponse( $result );
-		} );
-	}
-
-	private function parseGetPageResponse( $response ) {
-		$pages = $response['query']['pages'] ?? [];
-		foreach ( $pages as $page ) {
-			$title = $page['title'];
-			if ( isset( $page['revisions'] ) ) {
-				foreach ( $page['revisions'] as $revision ) {
-					return Util::getXhtmlFromContent( $this->getLang(), $revision['*'], $title );
```

Comment:
Why remove `Util::getXhtmlFromContent()`? This method is used to strip comments and turn the HTML into XHTML (as [required by epub](https://www.w3.org/publishing/epub32/epub-contentdocs.html#sec-xhtml-conf-content)). It also adds a reference to `main.css`.

Reply:
The new API responds with a page that includes the HTML already as follows: 
```
 <!DOCTYPE html>
<html prefix="dc: http://purl.org/dc/terms/ mw: http://mediawiki.org/rdf/" about="https://en.wikisource.org/wiki/Special:Redirect/revision/9909225"><head prefix="mwr: https://en.wikisource.org/wiki/Special:Redirect/"><meta property="mw:TimeUuid" content="b4fbd150-fa6e-11ea-aa5a-214d3c313213"/><meta charset="utf-8"/><meta property="mw:pageId" content="53313"/><meta property="mw:pageNamespace" content="0"/><link rel="dc:replaces" resource="mwr:revision/9909159"/><meta property="mw:revisionSHA1" content="bbc6e24f9721b0ec0b075a7c31391b4cab85ea96"/><meta property="dc:modified" content="2020-02-03T21:20:39.000Z"/><meta property="mw:html:version" content="2.1.0"/><link rel="dc:isVersionOf" href="//en.wikisource.org/wiki/Poems%2C_Consisting_Chiefly_of_Translations_from_the_Asiatick_Languages"/><title>Poems, Consisting Chiefly of Translations from the Asiatick Languages</title><base href="//en.wikisource.org/wiki/"/><link rel="stylesheet" href="/w/load.php?lang=en&amp;modules=mediawiki.skinning.content.parsoid%7Cmediawiki.skinning.interface%7Csite.styles%7Cmediawiki.page.gallery.styles%7Cext.cite.style%7Cext.cite.styles&amp;only=styles&amp;skin=vector"/><meta http-equiv="content-language" content="en"/><meta http-equiv="vary" content="Accept"/></head><body id="mwAA" lang="en" class="mw-content-ltr sitedir-ltr ltr mw-body-content parsoid-body mediawiki mw-parser-output" dir="ltr"><section...
```
So we wouldn't need to call the [Util::getXhtmlFromContent()](https://github.com/wsexport/tool/blob/d2d86f3df8901018a4921cfeee1470c2327cc796/src/Util/Util.php#L80) entirely since we now just need to remove comments and add `XML` so I added the logic in the [getPageAsync() function](https://github.com/wsexport/tool/compare/dev...migrate-to-parsoid-api#diff-0e7a3624f1374dc6effded1577ffd19342d9d23fcac98f0c45a98947fba7ce1bR193).

I didn't want to modify the `getXhtmlFromContent()` because it is being called in multiple places and the `$content` there might be retrieved from the old API.

The new API has a reference to different CSS. Do you think it needs to include the main.css reference?

Reply:
We need to produce XHTML for epubs, and there's also lots of other stuff in the head of the new content that we don't need. For example, we want something like this:

```html
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" dir="ltr">
<head>
	<meta content="application/xhtml+xml;charset=UTF-8" http-equiv="default-style"/>
	<link type="text/css" rel="stylesheet" href="main.css"/>
	<title>Emma</title>
</head>
```

but are getting:

```html
<!DOCTYPE html>
<html prefix="dc: http://purl.org/dc/terms/ mw: http://mediawiki.org/rdf/"
	  about="https://en.wikisource.org/wiki/Special:Redirect/revision/9471884">
<head prefix="mwr: https://en.wikisource.org/wiki/Special:Redirect/">
	<meta property="mw:TimeUuid" content="2793b590-fa54-11ea-973d-6144e17084ac"/>
	<meta charset="utf-8"/>
	<meta property="mw:pageId" content="3208"/>
	<meta property="mw:pageNamespace" content="0"/>
	<link rel="dc:replaces" resource="mwr:revision/9467089"/>
	<meta property="mw:revisionSHA1" content="164f861d4d643038c58877d4826a0085c9141d4f"/>
	<meta property="dc:modified" content="2019-07-23T15:52:33.000Z"/>
	<meta property="mw:html:version" content="2.1.0"/>
	<link rel="dc:isVersionOf" href="//en.wikisource.org/wiki/Emma"/>
	<title>Emma</title>
	<base href="//en.wikisource.org/wiki/"/>
	<link rel="stylesheet"
		  href="/w/load.php?lang=en&amp;modules=mediawiki.skinning.content.parsoid%7Cmediawiki.skinning.interface%7Csite.styles%7Cmediawiki.page.gallery.styles%7Cext.cite.style%7Cext.cite.styles&amp;only=styles&amp;skin=vector"/>
	<meta http-equiv="content-language" content="en"/>
	<meta http-equiv="vary" content="Accept"/>
</head>
```

The other two uses of `Util::getXhtmlFromContent()` are fixable: one is just wanting an empty document, and one is inserting a simple translation of the word "About" (which as far as I can see never contains any HTML, but I haven't checked all Wikisources).

Maybe it's worth encapsulating the changing of this HTML content here though, because it shouldn't ever be needed elsewhere. So it's fine if you want to keep it in `getPageAsync()`, but it should still use `getXhtmlFromContent()` because we still want the simplified `<head>`. Or refactor both to make it simpler, whatever you reckon is best.

The link to `main.css` is also still required, as that is the main CSS file for the epub (it's a concatenation of the font-family stuff if a font is requested, and the `MediaWiki:Epub.css` from the Wikisource).

Reply:
Thank you for the thorough explanation. I made the changes to [Util::getXhtmlFromContent() ](https://github.com/wsexport/tool/pull/267/files#diff-cdcb8e79143f7aa684522e60f70211cb3c6341043d1836c45e2ae1b61e30de30L80)so that it strips out the stuff before `<body` in the content retrieved from the new API endpoint.

Reply:
An other maybe cleaner option might be to drop completely this getXhtmlFromContent function and parse the content as HTML in PageParser (maybe using [Remex](https://www.mediawiki.org/wiki/RemexHtml)) and to the DOM changes to make it valid XHTML 5 inside of PageParser.

Clearing the Parsoid header removes also, I believe, the TemplateStyles data.

Reply:
That's a good idea. It'd be nice to have it all centralized in one place. Although, maybe it can be done as a second step; I'm not sure.

The template styles are moved to the head after this, by `PageParser::moveStyleToHead()`, I think. At least, it looks like it works with stripping the entire head at this point. :)